### ISSUE TITLE
Restore State failure

### ISSUE THREAD

**Issue by Vort** [bochs-emu/Bochs#96]
When I try to save and then restore state of machine, I encounter several problems:
1. Same as #95.
2. `ata0-0: could not open hard drive image file 'hd10meg.img'`
3. `cannot find param 'blk64'!`

I tested [dlxlinux4.tar.gz](https://bochs.sourceforge.io/guestos/dlxlinux4.tar.gz): started machine, clicked Suspend, selected directory, selected No, restarted Bochs, clicked Restore State and selected the same directory.

Version: 221cac79720387b33a84c915a922af3c9527f04e

**Comment by fysnet** [comment]
I could not get this to fail as you mention.

I load Bochs and let the Guest run a few seconds. Then I hit the Suspend button, choose a directory, then selected No to exit the application. I then used the `-r ./save' parameters to restore the Guest. The Guest was restored and continued execution from the point it left off.

One point though is that if I used the directory listing given by the Suspend button, to create a new directory, the first time I tried to save it, it failed to save. Simply hitting Suspend again and choosing the already existing directory succeeded. I did not try again to see if it would fail on a new directory creation. Something I should have done/should try again.

Also, the Restore button (under the Config menu) is grey'd out for me. What Restore button are you referring to?

If you can make this fail, please post your exact bochsrc.txt file again.

**Comment by Vort** [comment]
Please look at this video:

https://github.com/bochs-emu/Bochs/assets/1242858/30da96a8-28a8-4f7a-bc40-c36d08e18f79

**Comment by stlintel** [comment]
Don't you think it might be related to 'https://github.com/bochs-emu/Bochs/issues/37' ?

**Comment by Vort** [comment]
> Did you try to do 'Load' for loading configuration first ?

Just tried and Restore worked correctly this way.

But I think both ways should be functional.

**Comment by fysnet** [comment]
I see now how you are loading it. I use the -q parameter so that it automatically loads bochsrc. Without the -q parameter, I can recreate the error you are showing here, though mine stops at another config line item.

And yes, this might be related to #37.

**Comment by fysnet** [comment]
Okay, here is what I think.

When you load Bochs, before the dialog shows up to allow you to LOAD or RESTORE STATE, Bochs has *already* loaded the bochsrc.txt file in that directory. It had to, so it could get the configuration. For example, in this case, Win32.

`config_interface: win32config`

if it didn't it wouldn't know how to display the start dialog. Is it Windows, Linux, what?

However, it also ran through the whole bochsrc.txt file initializing all of the parameters. It shouldn't have since you haven't loaded or restored anything yet.

However, now when you go to restore state, some of the parameters have already been initialized and either cannot be re-initialized, or will throw an error since it has already been initialized.

In your case, I am thinking that the disk image has already been opened, and it is trying to reopen the handle and Bochs is giving an error because the handle is already open and protected from someone else opening the file.

Place the following line in main.cc at line 321:
`bx_reset_options();`
It should look like this:
```
      BX_INSTR_EXIT_ENV();
      return 0;
    }
    bx_reset_options();   // <------ this is the new line to add
    // read a param to decide which config interface to start.
    // If one exists, start it.  If not, just begin.
```
And then do exactly as you do in the video above.

Does it work correctly now? May not since the image file is still open.

This is just a hack. If you use the technique you show in the video, this might work. However, if you use the -q parameter to automagically load the bochsrc.txt file, it will fail miserably.

I believe the fix is if the -q parameter was not used, only parse the bochsrc file until you find the config_interface parameter, then quit parsing it, ignoring everything else, simply to get the configuration parameter.

Just my idea at the moment, further investigation will need to take place.

**Comment by fysnet** [comment]
> > When you load Bochs, before the dialog shows up to allow you to LOAD or RESTORE STATE, Bochs has already loaded the bochsrc.txt file in that directory. It had to, so it could get the configuration. For example, in this case, Win32.
> 
> Bochs can start from empty directory and I like this feature.

Absoluletly. I am not saying that it shouldn't. What I am saying is, that if you start Bochs and if it finds a bochsrc.txt file, it loads that bochsrc.txt file before it gives you the option to choose one with either the LOAD button or RESTORE STATE button.

i.e.: It has already loaded a bochsrc.txt file with all the options **before** you have a chance to tell it which file to load. Therefore, all of the options have already been loaded prematurely. These prematurely loaded options are messing with the new options.

For example, the USB config loading code will PANIC if you have already defined a controller's port options when trying to define the same port.  For example, the following will PANIC.
```
usb_uhci: port1=mouse, options1="speed:low, model:m338"
usb_uhci: port1=keyboard, options1="speed:low"
```
You are trying to define a new keyboard when you already defined a mouse on that port.

So, if Bochs prematurely loads another bochsrc.txt file that happens to have the following line:
`usb_uhci: port1=mouse, options1="speed:low, model:m338"`
then you chose to restore state from another bochrsc.txt file that happens to have the following line:
`usb_uhci: port1=keyboard, options1="speed:low"`
Bochs will PANIC since it already thinks there is a mouse on that port. However, the mouse option was prematurely loaded before you had a chance to click on the LOAD or RESTORE STATE buttons.

This is in error. Bochs should not initialize any parameters other than acquiring the `config_interface:` option before you have a chance to click on the LOAD or RESTORE STATE button.

In other words, before Bochs gets to the point of displaying the Win32 Dialog asking if you want to Load or Restore, it has found a bochsrc.txt file with the following line in it:
`config_interface: win32config`
If it didn't find a bochsrc.txt file, you would not be at the Win32 dialog I speak of, it would be using the text based configuration.

Unfortunately, Bochs also loaded all of the other options in that bochsrc.txt file and these options are interfering with the new options from the LOAD or RESTORE STATE procedure.

I don't know how to fix this at the moment. It looks to be quite the adventure to modify the loading procedure to fix this.

**Comment by Vort** [comment]
> If it didn't find a bochsrc.txt file, you would not be at the Win32 dialog I speak of, it would be using the text based configuration.

This is wrong.
I copied single `bochs.exe` file to VM which had no Bochs installed before. Result:
![bochs_w8](https://github.com/bochs-emu/Bochs/assets/1242858/eba2a456-a0f4-4b1c-b248-235f7c59ea3f)

**Comment by Vort** [comment]
I have figured out why `could not open hard drive image file` error appears.
It happens because of missing change of working directory.
Most likely, when config opening happens, Windows automatically change directory, but not when restore happens.
This change (hack-)fixes this problem:
```diff
diff --git a/bochs/gui/win32config.cc b/bochs/gui/win32config.cc
index 2d5413634..fb025fbff 100644
--- a/bochs/gui/win32config.cc
+++ b/bochs/gui/win32config.cc
@@ -603,6 +603,7 @@ static BOOL CALLBACK MainMenuDlgProc(HWND hDlg, UINT msg, WPARAM wParam, LPARAM
         case IDRESTORE:
           path[0] = 0;
           if (BrowseDir("Restore Bochs state from...", path) >= 0) {
+            SetCurrentDirectory(path);
             SIM->get_param_bool(BXPN_RESTORE_FLAG)->set(1);
             SIM->get_param_string(BXPN_RESTORE_PATH)->set(path);
             EndDialog(hDlg, 1); 
```
However, `cannot find param 'blk64'` do not disappears. Another bug chase is needed for it.

**Comment by stlintel** [comment]
The param blk64 is part of memory settings.
Total emulated ram is divided to 'pages' each block size is stored 'BX_MEM_THIS block_size'
All these details come from configure script - block size, number of blocks (memory size) and etc
You supposed to load them from somewhere (.bochsrc) before loading save/restore
In general if you saved 512MB ram with block size of 2MB you cannot restore it as 128MB or as block size 4MB

**Comment by Vort** [comment]
> You supposed to load them from somewhere (.bochsrc) before loading save/restore

This information is stored in the same place where disk image path is located: inside of `config` (which is different file from `bochsrc.txt`, but have similar meaning).
So question is why memory settings from `bochsrc.txt` can be loaded correctly and memory settings from `config` can't.

**Comment by stlintel** [comment]
What is `config `and why it is different from` .bochsrc` ?

**Comment by Vort** [comment]
> There is no such file as config  as far as I know

This file is created during state saving:
https://github.com/bochs-emu/Bochs/blob/325fb4bfe069ff5659c7473f6c2d708c0f3e3059/bochs/gui/siminterface.cc#L1082-L1083
And used during state restoring.

**Comment by Vort** [comment]
Looks like `cannot find param 'blk64'` appears because of reasons @fysnet described - conflict with some configuration file.
I can't say should it be fixed or not. Probably not. In such case this report can be closed.

**Comment by stlintel** [comment]
> > There is no such file as config  as far as I know
> 
> This file is created during state saving:
> 
> https://github.com/bochs-emu/Bochs/blob/325fb4bfe069ff5659c7473f6c2d708c0f3e3059/bochs/gui/siminterface.cc#L1082-L1083
> 
> 
> And used during state restoring.

The `write_rc` is traced to 

```
int bx_real_sim_c::write_rc(const char *rc, int overwrite)
{
  return bx_write_configuration(rc, overwrite);
}
```

where `bx_write_configuration` exactly writing to file current .bochsrc

The .bochsrc is the ONLY Bochs configuration
If exists - it is taken from there, if you restore an image -> it should be taken from restored config.
It looks like sometimes there is a mix or restored config is ignored/partially ignored.

**Comment by stlintel** [comment]
> Looks like `cannot find param 'blk64'` appears because of reasons @fysnet described - conflict with some configuration file. I can't say should it be fixed or not. Probably not. In such case this report can be closed.

As I said there ONLY one configuration file. And the conflict is not with "some" configuration file but with loaded and restored configuration. The loaded should take priority. Or ideally you should never supply both.

**Comment by Vort** [comment]
> The .bochsrc is the ONLY Bochs configuration

3 of them in my case.
1 is here: [.bochsrc](https://github.com/bochs-emu/Bochs/blob/325fb4bfe069ff5659c7473f6c2d708c0f3e3059/bochs/.bochsrc)
2 is `bochsrc.txt` from `dlxlinux4.tar.gz`.
3 is `config` made by state save.

> And the conflict is not with "some" configuration file but with loaded and restored configuration.

In my case, between number 1 and number 3. Number 2 is ignored during state restore.

> Or ideally you should never supply both.

Number 1 is shipped with source code. And build scripts produce `.exe` file right next to it.
If I move `.exe` somewhere, `'blk64'` problem disappear.

**Comment by Vort** [comment]
Conflict happens because of `block_size` parameter inside of `.bochsrc` file:
https://github.com/bochs-emu/Bochs/blob/325fb4bfe069ff5659c7473f6c2d708c0f3e3059/bochs/.bochsrc#L372
If I add `, block_size=128` manually into `config`, then `'blk64'` problem disappear.

> To restore I do Bochs -r restore_image

I tested it: change of current directory should be added to this code path too.

**Comment by stlintel** [comment]
I pushed a fix to add block_size to config
Can you check and report if the issue disppeared ?

**Comment by Vort** [comment]
> Can you check and report if the issue disappeared ?

I confirm that `'blk64'` problem disappeared.

But I think problem with `-r` needs to be fixed too.

**Comment by stlintel** [comment]
> > Can you check and report if the issue disappeared ?
> 
> I confirm that `'blk64'` problem disappeared.
> 
> But I think problem with `-r` needs to be fixed too.

I believe it could deserve a separate record with clear description and instructions to reproduce the issue.
Closing this one for now as resolved.
